files/default/neutron-ha-tool.py

```diff
@@ -576,6 +600,195 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
     return errors
 
 
+def dhcp_agent_evacuate(qclient, agent_host, agent_picker,
+                      noop=False, wait_for_network=True, fail_fast=False):
+    """
+    Retreive a list of networks scheduled on the listed DHCP agent, and move that
+    to another agent.
+
+    :param qclient: A neutronclient
+    :param noop: Optional noop flag
+    :param agent_host: the hostname of the DHCP agent to migrate networks from
+    :returns: total number of errors encountered
+
+    """
+
+    agent_list = list_agents(qclient, agent_type='DHCP agent')
+
+    agent_to_evacuate = None
+    for agent in agent_list:
+        if agent.get('host', None) == agent_host:
+            agent_to_evacuate = agent
+            break
+
+    if not agent_to_evacuate:
+        LOG.error("Could not locate agent to evacuate; aborting!")
+        return 0
+
+    target_list = list_alive_dhcp_agents_except(agent_list,
+                                                exclude_agent=agent_to_evacuate)
+
+    if len(target_list) < 1:
+        LOG.error("There are no DHCP agents alive to migrate routers onto")
+        return 0
+
+    (migrations, errors) = \
+        migrate_dhcp_networks_from_agent(qclient, agent_to_evacuate,
+                                      target_list, agent_picker,
+                                      noop, wait_for_network, fail_fast)
+    LOG.info("%d networks %s evacuated from DHCP agent %s", migrations,
+             "would have been" if noop else "were", agent_host)
+    if errors > 0:
+        LOG.error("%d errors encountered during evacuation", errors)
+
+    return errors
+
+
+def migrate_dhcp_networks_from_agent(qclient, agent, targets, agent_picker,
+                                     noop, wait_for_network, fail_fast):
+    LOG.info("Querying agent_id=%s for networks to migrate away", agent['id'])
+
+    networks_on_agent = qclient.list_networks_on_dhcp_agent(agent['id'])['networks']
+
+    agent_picker.set_agents(targets)
+
+    migrations = 0
+    errors = 0
+    for network in networks_on_agent:
+        target = agent_picker.pick()
+
+        if noop:
+            LOG.info("Would try to migrate network=%s from agent=%s "
+                 "to agent=%s", network['id'], agent['id'], target['id'])
+            migrations += 1
+            continue
+
+        try:
+            migrate_dhcp_network(qclient, network, agent, target, wait_for_network)
+            migrations += 1
+        except:
+            LOG.exception("Failed to migrate network_id=%s from agent=%s to "
+                          "agent=%s", network['id'], agent['id'], target['id'])
+            errors += 1
+            if fail_fast:
+                break
+
+    return (migrations, errors)
+
+
+def migrate_dhcp_network(qclient, network, agent, target,
+                         wait_for_network=True):
+    """
+    Returns nothing, and raises exceptions on errors.
```

Comment:
That might be a good idea ... but for example `migrate_router_safely` is not using exceptions but return values.

It could be a good thing to have it consistent in the same tool...

Reply:
I'm not really sure what is the benefit of this additional "safely" wrapper function. The only additional functionality is skipping migration of live agents which is not included in DHCP evacuation. Besides that, I removed this additional nesting level and included the try-catch directly in `migrate_dhcp_networks_from_agent`.

Reply:
I'm just pointing out to the different way dhcp_ and router_ methods are reporting errors. It would be nicer if ti were consistent

Reply:
> I'm not really sure what is the benefit of this additional "safely" wrapper function. 

Me neither. But I assume it was done for readablity reason. See https://github.com/SUSE-Cloud/cookbook-openstack-network/commit/41a048c5325c62da8207adcfd5bc531ee57b323a

Not sure if the dhcp functions reached that state for complexity yet.